### PR TITLE
Reformatted scLinac/updated Rack naming convention. Also changed archiver data handling

### PR DESCRIPTION
**Pull request by lisazacarias**

### CODE REVIEW

File: lcls_tools/data_analysis/archiver.py [hunk 1]
```diff
@@ -18,82 +20,97 @@
 TIMEOUT = 3
 
 
+@dataclass()
+class ArchiverData:
+    # todo make this not a union
```
Comment (nneveu) [inline lcls_tools/data_analysis/archiver.py:25]: Can you add some comments here on the main goal of this data class? I think it's fairly obvious, but wouldn't hurt to write it down.

File: lcls_tools/data_analysis/archiver.py [hunk 2]
```diff
@@ -18,82 +20,97 @@
...
+    # Currently one list if timestamps are shared by all PVs else a dict of
+    # pv -> timstamps
+    timeStamps: Union[List[datetime], Dict[str, List[datetime]]]
+    values: Dict[str, List]
+
+
 class Archiver(object):
     def __init__(self, machine):
```
Comment (nneveu) [inline lcls_tools/data_analysis/archiver.py:33]: Same here, add a comment or two on inputs/outputs for the class and use case? i.e. what does machine look like?

File: lcls_tools/data_analysis/archiver.py [hunk 3]
```diff
@@ -18,82 +20,97 @@
...
+        values = {}
+        times = [timeRequested]
+        for pv in pvList:
+            values[pv] = []
 
         try:
             jsonData = json.loads(response.text)
-            for key, val in jsonData.items():
-                results[key] = val[u'val']
+            for pv, data in jsonData.items():
```
Comment (nneveu) [inline lcls_tools/data_analysis/archiver.py:53]: nice name changes - > pv etc, makes the code more clear!

File: lcls_tools/data_analysis/archiver.py [hunk 4]
```diff
@@ -18,82 +20,97 @@
...
-            results[pv] = {"times": [], "values": []}
+            values[pv] = []
 
         while currTime < endTime:
             result = self.getDataAtTime(pvList, currTime)
-            for key, val in result.items():
-                results[key]["times"].append(currTime)
-                results[key]["values"].append(val)
+            times.append(currTime)
+            for key, val in result.values.items():
```
Comment (nneveu) [inline lcls_tools/data_analysis/archiver.py:72]: could this key also be 'pv'?

File: lcls_tools/data_analysis/archiver.py [hunk 5]
```diff
@@ -18,82 +20,97 @@
...
             for pv in pvList:
+                times[pv] = []
+                values[pv] = []
 
                 response = requests.get(url=url, timeout=TIMEOUT,
-                                        params={"pv": pv,
+                                        params={"pv"  : pv,
                                                 "from": startTime.isoformat() + "-07:00",
-                                                "to": endTime.isoformat() + "-07:00"})
+                                                "to"  : endTime.isoformat() + "-07:00"})
```
Comment (nneveu) [inline lcls_tools/data_analysis/archiver.py:99]: Maybe add a comment here about the -7 offset for people who don't know why we need that.

File: lcls_tools/data_analysis/archiver.py [hunk 6]
```diff
@@ -18,82 +20,97 @@
...
-                    result = {"times": [], "values": []}
                     for datum in element[u'data']:
-                        result["times"].append(datum[u'secs'])
-                        result["values"].append(datum[u'val'])
-
-                    results[pv] = result
+                        # TODO implement filtering by BSA PV
+                        times[pv].append(datetime.fromtimestamp(datum[u'secs'])
+                                         + timedelta(microseconds=datum[u'nanos'] / 1000))
+                        values[pv].append(datum[u'val'])
```
Comment (nneveu) [inline lcls_tools/data_analysis/archiver.py:109]: Are we confident these keys will stay the same? I'm unfamiliar. Can we pull them into a variable that would update if the key changed?

File: lcls_tools/data_analysis/archiver.py [hunk 7]
```diff
@@ -108,49 +125,193 @@ def __init__(self, *args, **kwargs):
...
+                     9.338020621211182, 9.338010311844357, 9.338005157161007,
+                     9.338002579819346, 9.357603822477754, 10.709958700542789,
+                     10.709979350271261, 10.7099896751356, 10.690381982944237,
+                     9.33804124242413, 9.33804124242413, 10.709958699869212,
+                     10.709979349934471, 10.7099896749672]
+
+        self.expectedNoDeltaResult = ArchiverData(timeStamps={'BEND:LTUH:280:BDES': times280,
+                                                              'BEND:LTUH:220:BDES': times220},
+                                                  values={'BEND:LTUH:280:BDES': values280,
+                                                          'BEND:LTUH:220:BDES': values220})
```
Comment (nneveu) [inline lcls_tools/data_analysis/archiver.py:315]: I love that you have a test, so good!

File: lcls_tools/data_analysis/archiver.py [hunk 8]
```diff
@@ -108,49 +125,193 @@ def __init__(self, *args, **kwargs):
...
 
     class MockResponse(object):
         def __init__(self):
```
Comment (nneveu) [inline lcls_tools/data_analysis/archiver.py:317]: Add some comments here on what this class is for?

File: lcls_tools/data_analysis/archiver.py [hunk 9]
```diff
@@ -191,35 +352,45 @@ def side_effect(pvList, time):
...
+                return ArchiverData(timeStamps=[datetime(2020, 10, 12, 11, 56, 30, 10)],
+                                    values={u'BEND:LTUH:280:BDES': [10.709999116032312],
+                                            u'BEND:LTUH:220:BDES': [10.709999125516292]})
             elif time == datetime(2020, 10, 13, 11, 56, 30, 10):
-                return {u'BEND:LTUH:280:BDES': 10.70997130324195,
-                        u'BEND:LTUH:220:BDES': 10.709944052743765}
+                return ArchiverData(timeStamps=[datetime(2020, 10, 13, 11, 56, 30, 10)],
+                                    values={u'BEND:LTUH:280:BDES': [10.70997130324195],
+                                            u'BEND:LTUH:220:BDES': [10.709944052743765]})
             else:
```
Comment (nneveu) [inline lcls_tools/data_analysis/archiver.py:394]: Why the switch/addition of this time format  'datetime(2020, 10, 13, 11, 56, 30, 10)' in these cases? Was 10.70997130324195 from before not enough?

**Review by nneveu — APPROVED** [review]
I left some general comments/questions, but I approve for the merge. You can do another commit if you want to include any of the suggestions I made.

File: lcls_tools/data_analysis/archiver.py (hunk 6 above)
Comment (lisazacarias) [reply lcls_tools/data_analysis/archiver.py:109]: Honestly I have no idea I'm just following the [documentation](https://slacmshankar.github.io/epicsarchiver_docs/userguide.html) I found online, but I can at least link it in some comments

**Review by picarro-aegger — COMMENTED** [review]
This is interesting.  I'd like to learn more about what the intention is here.  Also, how are the envs being setup, Conda?  For some reason I linked my email to this GitHub repo, so now I'm invested!

I'm really liking that this is looking like py3 with nice type hinting and using dataclass.  You should look at pydantic, I use it for everything.

Also, I'm completely shocked this repo is still in existence, lol

File: lcls_tools/data_analysis/archiver.py [hunk 10]
```diff
@@ -18,82 +20,105 @@
 TIMEOUT = 3
 
 
+# Adding a data class for easier data handling (so that external functions can
+# invoke .values and .timestamps instead of having to remember dictionary keys
+@dataclass()
```
Comment (picarro-aegger) [inline lcls_tools/data_analysis/archiver.py:25]:
```suggestion
@dataclass
```

File: lcls_tools/data_analysis/archiver.py [hunk 11]
```diff
@@ -18,82 +20,105 @@
...
+class ArchiverData:
+    # todo make this not a union
+    # Currently one list if timestamps are shared by all PVs else a dict of
+    # pv -> timstamps
+    timeStamps: Union[List[datetime], Dict[str, List[datetime]]]
+    values: Dict[str, List]
+
+
 class Archiver(object):
```
Comment (picarro-aegger) [inline lcls_tools/data_analysis/archiver.py:34]:
```suggestion
class Archiver:
```

File: lcls_tools/data_analysis/archiver.py [hunk 12]
```diff
@@ -18,82 +20,105 @@
...
+
+    # machine is a string that is either "lcls" or "facet"
     def __init__(self, machine):
         # type: (str) -> None
         self.url_formatter = ARCHIVER_URL_FORMATTER.format(MACHINE=machine)
 
     def getDataAtTime(self, pvList, timeRequested):
```
Comment (picarro-aegger) [inline lcls_tools/data_analysis/archiver.py:41]:
```suggestion
    def get_data_at_time(self, pv_list: List[str], time_requested: datetime):
```
Depending on what python version you're using.  Main take away is use snake case for all things python.

Anyway, bouncing out here, but I'd love to help you finish this.  

Rules:
Never leave a PR open for more than 2 weeks... even at a national lab
Quick iteration is best
Write tests
No magic numbers and no hard coding
If you go in on type hinting/validation, go all in

Comment (picarro-aegger) [reply lcls_tools/data_analysis/archiver.py:41]: I forgot to say that I hope all is well at SLAC.  I'm super happy to see some great people making great things happen!

File: lcls_tools/data_analysis/archiver.py (hunk 9 above)
Comment (nneveu) [reply lcls_tools/data_analysis/archiver.py:403]: No the elif doesn't bother me, I was wondering about the format for times here. Did the format change or am I misunderstanding this?

**Pull request merged**